### ISSUE TITLE
the codebase design has high coupling and need to be refactored

### ISSUE THREAD

**Issue by walkacross** [opendilab/DI-engine#666]
The advantage is that the algorithm in the project is very rich, and the disadvantage is that the code coupling of the entire project is too high and there is no modular design in fact. 

The disadvantage of high coupling is that few users can really use some of the functions or class, if you really hope users use the library you create, refactot it. 

Think and learn why libraries with high quality design, e.g. pytorch, hugging face,  success.

**Comment by PaParaZz1** [comment]
Thanks for your comments. Do you have any detailed suggestions for us to improve further, such as some specific modules or functions?

**Comment by walkacross** [comment]
hi @PaParaZz1, thank for your response. my personal suggestion are as follows, just for reference.

## background

the [DI-engine](https://github.com/opendilab/DI-engine) combines so many different functionalities into one library, e.g. env or its wrappers, buffer, algorithms, training process et al, which results in high coupling and low readability and resusability. It seems work well as a whole, but the single module has low generalization and few users can really use this library, even it is hard to use some of the functions or class in their research pipeline.

## suggestions
Just like hugging face does, we should split the whole project into different libraries where every single library has only one main research purpose, following the Single Responsibility Principle.

* rl model library and its pretrained params.  like the positioning of hugging face [transformer](https://github.com/huggingface/transformers) library.
* a general rl trainging framework. Corresponding to hugging face [accelerate](https://github.com/huggingface/accelerate) library positioning.
* a general Experience Buffer library, handling priority/non-priority, sequence/none-sequence experience. just like [tokenizer](https://github.com/huggingface/tokenizers) positioning.
*  a model-base rl library, handling model-based case. just like [pytorch-image-models](https://github.com/huggingface/pytorch-image-models), [diffusers](https://github.com/huggingface/diffusers) positioning.

et al.

**Comment by PaParaZz1** [comment]
Thanks for your suggestions, we are planning to separate `reward model` and `world model` parts for different sub domains. For the design of replay buffer, we have already refactored this part with new storage and middleware design, you can refer to this [doc](https://di-engine-docs.readthedocs.io/en/latest/04_best_practice/buffer.html). For pretrained parameters, we are working on Hugging Face + DI-engine Model Zoo, you can refer to [here](https://huggingface.co/OpenDILabCommunity) for more details.
